ksqldb-engine/src/main/java/io/confluent/ksql/planner/LogicalPlanner.java

```diff
@@ -1092,6 +1106,48 @@ private LogicalSchema buildRepartitionedSchema(
     );
   }
 
+  /**
+   * Given a source node and its side, return {@code true} if the source should be repartitioned
+   * for the foreign key join.
+   *
+   * <p>The right hand side of the FK join must be repartitioned if it uses schema format baked by
+   * Schema Registry and the primary key is a struct. In that case, there is a possibility that
+   * the schema was not created by ksqlDB, therefore it won't match the generated without
+   * repartitioning. See https://github.com/confluentinc/ksql/issues/8528 for more details</p>
+   *
+   * @param sourceNode the source node participating in a FK join
+   * @param side the side of the FK join the source node participates at.
+   * @return {@code true} if the source node requires repartitioning, {@code false} otherwise.
+   */
+  private boolean shouldRepartitionFKJoin(final PlanNode sourceNode, final JoinSide side) {
+    if (!(sourceNode instanceof DataSourceNode)) {
+      return false;
+    }
+
+    final DataSourceNode dataSourceNode = (DataSourceNode) sourceNode;
+    final boolean isAvro = dataSourceNode.getDataSource()
+        .getKsqlTopic()
+        .getKeyFormat()
+        .getFormat()
+        .equals(AvroFormat.NAME);
+    final boolean isProtobuf = dataSourceNode.getDataSource()
+        .getKsqlTopic()
+        .getKeyFormat()
+        .getFormat()
+        .equals(ProtobufFormat.NAME);
+    final boolean isJsonSR = dataSourceNode.getDataSource()
+        .getKsqlTopic()
+        .getKeyFormat()
+        .getFormat()
+        .equals(JsonSchemaFormat.NAME);
+    final boolean isSREnabledFormat = isAvro || isProtobuf || isJsonSR;
+    final boolean hasStructInPrimaryKey = dataSourceNode.getSchema()
+        .key()
+        .stream()
+        .anyMatch(col -> col.type() instanceof SqlStruct);
```

Comment:
`SqlStruct` is really `STRUCT`, right? IIRC we only support Joins on primitive types though?

Reply:
Yeah, `SqlStruct` is the implementation of `SqlType` that maps to connect's `Struct`. See the converter:

https://github.com/confluentinc/ksql/blob/aeae363c4c2d265241aa30583e121be01a81d5c0/ksqldb-engine-common/src/main/java/io/confluent/ksql/schema/ksql/DefaultSqlValueCoercer.java#L245-L276

I didn't find any restrictions on FK or regular joins that prevent using non-primitive types.